File: apps/android/core/crypto/src/main/java/app/ollo/crypto/EnvelopePlanner.kt
package app.ollo.crypto

/**
 * Offline-first outbound planner. Must stay in lockstep with
 * `packages/shared/src/outbox.ts`. Native production encrypt still uses
 * libsignal; this only decides *whether* to consume a one-time prekey.
 */
object EnvelopePlanner {
    const val MAX_ATTEMPTS = 8
    const val BASE_DELAY_MS = 1500L

    enum class Status {
        Draft,
        Pending,
        Encrypted,
        Uploading,
        Sent,
        Delivered,
        Read,
        Failed,
        Retrying,
    }

    enum class KeyPlan {
        SkipSelf,
        UseSession,
        ConsumeBundle,
    }

    data class OutboxItem(
        val id: String,
        val status: Status,
        val attempts: Int,
    )

    fun nextRetryDelayMs(attempts: Int): Long {
        val exp = attempts.coerceIn(0, 6)
        return BASE_DELAY_MS * (1L shl exp)
    }

    fun onSendFailure(item: OutboxItem): OutboxItem {
        val attempts = item.attempts + 1
        return if (attempts >= MAX_ATTEMPTS) {
            item.copy(attempts = attempts, status = Status.Failed)
        } else {
            item.copy(attempts = attempts, status = Status.Retrying)
        }
    }

    /**
     * Never consume a one-time prekey when a Double Ratchet session exists.
     * Never address the sending device.
     */
    fun planKeyFetch(
        localUserId: String,
        localDeviceId: String,
        targetUserId: String,
        targetDeviceId: String,
        hasSession: Boolean,
    ): KeyPlan {
        if (localUserId == targetUserId && localDeviceId == targetDeviceId) {
            return KeyPlan.SkipSelf
        }
        return if (hasSession) KeyPlan.UseSession else KeyPlan.ConsumeBundle
    }

    const val PREKEY_MIN_DEPTH = 20
    const val PREKEY_BATCH = 100

    data class Replenish(val count: Int, val startId: Int)

    fun planPrekeyReplenish(remaining: Int, nextId: Int): Replenish? {
        if (remaining >= PREKEY_MIN_DEPTH) return null
        if (nextId < 1) return null
        return Replenish(count = PREKEY_BATCH, startId = nextId)
    }

    fun onRefreshRejected(): AuthFailure = AuthFailure.Wipe

    fun afterUnauthorized(refreshSucceeded: Boolean): AuthFailure =
        if (refreshSucceeded) AuthFailure.Retry else AuthFailure.Wipe

    enum class AuthFailure { Wipe, Retry }

    enum class SessionLaunch { SignedIn, NeedAuth }

    /** A wrapped vault session skips OTP. Missing secrets require a bound engine. */
    fun planSessionLaunch(hasVaultSession: Boolean): SessionLaunch =
        if (hasVaultSession) SessionLaunch.SignedIn else SessionLaunch.NeedAuth

    const val SIGNED_PREKEY_MAX_AGE_MS = 7L * 24 * 60 * 60 * 1000
    const val SIGNED_PREKEY_KEEP = 2

    fun keepSignedPrekeyIds(currentId: Int, storedIds: List<Int>): List<Int> {
        val retired = storedIds.filter { it != currentId && it > 0 }.sortedDescending()
        val keep = linkedSetOf<Int>()
        if (currentId > 0) keep.add(currentId)
        keep.addAll(retired.take(SIGNED_PREKEY_KEEP))
        return keep.toList()
    }

    data class SignedPrekeyPlan(val nextId: Int)

    fun planRosterPrune(sessionKeys: List<String>, userId: String, liveDeviceIds: Collection<String>): List<String> {
        if (userId.isEmpty()) return emptyList()
        val live = liveDeviceIds.toSet()
        val prefix = "$userId:"
        return sessionKeys.filter { it.startsWith(prefix) && it.removePrefix(prefix) !in live }
    }

    fun planDeviceDrop(sessionKeys: List<String>, userId: String, deviceId: String): List<String> {
        if (userId.isEmpty() || deviceId.isEmpty()) return emptyList()
        val target = "$userId:$deviceId"
        return sessionKeys.filter { it == target }
    }

    fun planSignedPrekeyRotation(currentId: Int, createdAtMs: Long?, now: Long): SignedPrekeyPlan? {
        if (currentId < 1) return null
        if (createdAtMs == null) return null
        if (now - createdAtMs < SIGNED_PREKEY_MAX_AGE_MS) return null
        return SignedPrekeyPlan(nextId = currentId + 1)
    }

    const val REPLAY_CACHE_MAX = 4096

    enum class ReplayDecision { Accept, Drop }

    /** Bounded FIFO of seen envelope ids. Duplicates must not re-apply. */
    fun rememberEnvelope(ids: MutableList<String>, envelopeId: String, max: Int = REPLAY_CACHE_MAX): ReplayDecision {
        if (envelopeId.isEmpty() || max < 1) return ReplayDecision.Drop
        if (ids.contains(envelopeId)) return ReplayDecision.Drop
        ids.add(envelopeId)
        while (ids.size > max) ids.removeAt(0)
        return ReplayDecision.Accept
    }
}

File: apps/ios/Sources/OlloCrypto/EnvelopePlanner.swift
import Foundation

/// Offline-first outbound planner. Must stay in lockstep with
/// `packages/shared/src/outbox.ts`. Production encrypt still uses libsignal;
/// this only decides whether to consume a one-time prekey.
public enum EnvelopePlanner {
    public static let maxAttempts = 8
    public static let baseDelayMs: Int = 1500

    public enum Status: String, Sendable {
        case draft
        case pending
        case encrypted
        case uploading
        case sent
        case delivered
        case read
        case failed
        case retrying
    }

    public enum KeyPlan: String, Sendable {
        case skipSelf = "skip-self"
        case useSession = "use-session"
        case consumeBundle = "consume-bundle"
    }

    public struct OutboxItem: Sendable {
        public var id: String
        public var status: Status
        public var attempts: Int

        public init(id: String, status: Status, attempts: Int) {
            self.id = id
            self.status = status
            self.attempts = attempts
        }
    }

    public static func nextRetryDelayMs(attempts: Int) -> Int {
        let exp = min(6, max(0, attempts))
        return baseDelayMs * (1 << exp)
    }

    public static func onSendFailure(_ item: OutboxItem) -> OutboxItem {
        var next = item
        next.attempts += 1
        next.status = next.attempts >= maxAttempts ? .failed : .retrying
        return next
    }

    /// Never consume a one-time prekey when a Double Ratchet session exists.
    /// Never address the sending device.
    public static func planKeyFetch(
        localUserId: String,
        localDeviceId: String,
        targetUserId: String,
        targetDeviceId: String,
        hasSession: Bool
    ) -> KeyPlan {
        if localUserId == targetUserId && localDeviceId == targetDeviceId {
            return .skipSelf
        }
        return hasSession ? .useSession : .consumeBundle
    }

    public static let prekeyMinDepth = 20
    public static let prekeyBatch = 100

    public struct Replenish: Sendable, Equatable {
        public var count: Int
        public var startId: Int
    }

    public static func planPrekeyReplenish(remaining: Int, nextId: Int) -> Replenish? {
        if remaining >= prekeyMinDepth { return nil }
        if nextId < 1 { return nil }
        return Replenish(count: prekeyBatch, startId: nextId)
    }

    public enum AuthFailure: String, Sendable {
        case wipe
        case retry
    }

    public static func onRefreshRejected() -> AuthFailure { .wipe }

    public static func afterUnauthorized(refreshSucceeded: Bool) -> AuthFailure {
        refreshSucceeded ? .retry : .wipe
    }

    public enum SessionLaunch: String, Sendable {
        case signedIn = "signed-in"
        case needAuth = "need-auth"
    }

    /// A wrapped vault session skips OTP. Missing secrets require a bound engine.
    public static func planSessionLaunch(hasVaultSession: Bool) -> SessionLaunch {
        hasVaultSession ? .signedIn : .needAuth
    }

    public static let signedPrekeyMaxAgeMs: Int64 = 7 * 24 * 60 * 60 * 1000
    public static let signedPrekeyKeep = 2

    public static func keepSignedPrekeyIds(currentId: Int, storedIds: [Int]) -> [Int] {
        let retired = storedIds.filter { $0 != currentId && $0 > 0 }.sorted(by: >)
        var keep: [Int] = []
        if currentId > 0 { keep.append(currentId) }
        for id in retired.prefix(signedPrekeyKeep) where !keep.contains(id) {
            keep.append(id)
        }
        return keep
    }

    public struct SignedPrekeyPlan: Sendable, Equatable {
        public var nextId: Int
    }

    public static func planRosterPrune(sessionKeys: [String], userId: String, liveDeviceIds: [String]) -> [String] {
        if userId.isEmpty { return [] }
        let live = Set(liveDeviceIds)
        let prefix = "\(userId):"
        return sessionKeys.filter { key in
            guard key.hasPrefix(prefix) else { return false }
            return !live.contains(String(key.dropFirst(prefix.count)))
        }
    }

    public static func planDeviceDrop(sessionKeys: [String], userId: String, deviceId: String) -> [String] {
        if userId.isEmpty || deviceId.isEmpty { return [] }
        let target = "\(userId):\(deviceId)"
        return sessionKeys.filter { $0 == target }
    }

    public static func planSignedPrekeyRotation(currentId: Int, createdAtMs: Int64?, now: Int64) -> SignedPrekeyPlan? {
        if currentId < 1 { return nil }
        guard let createdAtMs else { return nil }
        if now - createdAtMs < signedPrekeyMaxAgeMs { return nil }
        return SignedPrekeyPlan(nextId: currentId + 1)
    }

    public static let replayCacheMax = 4096

    public enum ReplayDecision: String, Sendable {
        case accept
        case drop
    }

    /// Bounded FIFO of seen envelope ids. Duplicates must not re-apply.
    @discardableResult
    public static func rememberEnvelope(_ ids: inout [String], envelopeId: String, max: Int = replayCacheMax) -> ReplayDecision {
        if envelopeId.isEmpty || max < 1 { return .drop }
        if ids.contains(envelopeId) { return .drop }
        ids.append(envelopeId)
        if ids.count > max {
            ids.removeFirst(ids.count - max)
        }
        return .accept
    }
}
